contracts/eosio.system/src/voting.cpp

```diff
@@ -110,13 +110,9 @@ namespace eosiosystem {
 
       using value_type = std::pair<eosio::producer_authority, uint16_t>;
       std::vector< value_type > top_producers;
-      std::vector< eosio::finalizer_authority > next_finalizer_authorities;
-      std::vector< uint64_t > new_finalizer_key_ids;
-      std::unordered_set< uint64_t > kept_finalizer_key_ids;
+      std::vector< proposed_finalizer_key_t > proposed_fin_keys;
       top_producers.reserve(21);
-      next_finalizer_authorities.reserve(21);
-      new_finalizer_key_ids.reserve(21);
-      bool new_finalizer_keys_found = false;
+      proposed_fin_keys.reserve(21);
 
       for( auto it = idx.cbegin(); it != idx.cend() && top_producers.size() < 21 && 0 < it->total_votes && it->active(); ++it ) {
          if( is_savanna_consensus() ) {
```

Comment:
To avoid the possibility that `get_last_proposed_finalizers` (which would replace `is_savanna_consensus()` here) could be expensive (especially if you avoid the singleton pattern), you could make a single call to `get_last_proposed_finalizers` near the top of `update_elected_producers` and set the returned vector to `last_proposed_finalizers`.

Then you can define the boolean:
```cpp
bool is_savanna_consensus = !last_proposed_finalizers.empty();`
```

That way you can cheaply use it within this for loop.